src/elm/Data/Event.elm

```diff
@@ -0,0 +1,32 @@
+module Data.Event exposing (Event(..), random, optionalRandom)
+
+import Data.Resource as Resource
+import Random exposing (Generator)
+import Random.Extra as Random
+
+
+type Event
+    = GlobalRateIncrease
+    | LocalRateIncrease Resource.Level
+
+
+all : List Event
+all =
+    [ GlobalRateIncrease
+    , LocalRateIncrease Resource.L1
+    , LocalRateIncrease Resource.L2
+    , LocalRateIncrease Resource.L3
+    , LocalRateIncrease Resource.L4
+    ]
+
+
+optionalRandom : Generator (Maybe Event)
+optionalRandom =
+    random
+        |> Random.maybe (Random.oneIn 2)
+
+
+random : Generator Event
+random =
+    Random.sample all
+        |> Random.map (Maybe.withDefault GlobalRateIncrease)
```

Comment:
If I don't have a default, then the generator will return a `Maybe Event` rather than an `Event`. because it's possible to pass in an empty list to `Random.sample`.

Reply:
Got it, makes sense.  

Just out of curiosity, could it be ```Random.map (Maybe.withDefault Nothing)```?  I ask because it seems strange to have a default for something that will never use it, since ```all``` here will always be not empty.

Reply:
`Maybe.withDefault` is the equivalent of:

```elm
case maybe of
  Just value -> value
  Nothing -> default
```

Since the value is already a `Maybe` it doesn't make sense to default it to `Nothing`.

It is frustrating to add a default here since _I_ know that `all` is not empty. The problem is that the _compiler_ can't prove that.

It's like doing:

```elm
List.head [1,2,3]
-- returns `Just 1`
```

It's annoying to get a `Maybe` back because _I_ know that the list will never be empty but the compiler can't prove it.

The classic way to get around that is to use a non-empty list. There are [packages](http://package.elm-lang.org/packages/mgold/elm-nonempty-list/3.1.0/List-Nonempty) for that but you can also do it yourself by requiring a first value with types like

```elm
type NoneEmptyList a = NonEmptyList a (List a)

-- or

type alias NonEmptyList a = { head : a, rest : List a }
````

You can do a similar thing with functions by requiring two parameters. You _could_ write a `Random.sample` function with a signature like:

```elm
sample : a -> List a -> Generator a
```

This approach is being taken for the next version of Elm. You can see a discussion of it in the [alpha docs](https://alpha.elm-lang.org/packages/elm/random/latest/Random#uniform).

Reply:
@JoelQ Great explanation, thank you!